haskell/cabal.bzl

```diff
@@ -0,0 +1,247 @@
+"""Cabal packages"""
+
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("@bazel_tools//tools/cpp:toolchain_utils.bzl", "find_cpp_toolchain")
+load(":cc.bzl", "cc_interop_info")
+load(":private/context.bzl", "haskell_context")
+load(":private/dependencies.bzl", "gather_dep_info")
+load(":private/mode.bzl", "is_profiling_enabled")
+load(":private/set.bzl", "set")
+load(
+    ":providers.bzl",
+    "HaskellInfo",
+    "HaskellLibraryInfo",
+    "empty_HaskellCcInfo",
+    "get_libs_for_ghc_linker",
+    "merge_HaskellCcInfo",
+)
+
+def _execute_or_fail_loudly(repository_ctx, arguments):
+    exec_result = repository_ctx.execute(arguments)
+    if exec_result.return_code != 0:
+        fail("\n".join(["Command failed: " + " ".join(arguments), exec_result.stderr]))
+    return exec_result
+
+def _so_extension(hs):
+    return "dylib" if hs.toolchain.is_darwin else "so"
+
+def _dirname(file):
+    return file.dirname
+
+def _haskell_cabal_library_impl(ctx):
+    hs = haskell_context(ctx)
+    dep_info = gather_dep_info(ctx, ctx.attr.deps)
+    cc = cc_interop_info(ctx)
+    cc_info = cc_common.merge_cc_infos(
+        cc_infos = [dep[CcInfo] for dep in ctx.attr.deps if CcInfo in dep],
+    )
+    name = hs.label.name
+    with_profiling = is_profiling_enabled(hs)
+
+    # Check that a .cabal file exists. Choose the root one.
+    cabal = None
+    for f in ctx.files.srcs:
+        if f.extension == "cabal":
+            if not cabal or f.dirname < cabal.dirname:
+                cabal = f
+    if not cabal:
+        fail("A .cabal file was not found in the srcs attribute.")
+
+    # Check that a Setup script exists. If not, create a default one.
+    setup = None
+    for f in ctx.files.srcs:
+        if f.basename in ["Setup.hs", "Setup.lhs"]:
+            if not setup or f.dirname < setup.dirname:
+                setup = f
```

Comment:
Same for the setup script, it should always be on the toplevel if it exists. Otherwise we might run into packages that have multiple setup files because they use the cabal project stuff.

Reply:
And there’s no breakage because of cabal projects with multiple `Setup.hs`?

Reply:
Cabal projects are not in source distributions. This deals with source distributions, i.e. single Cabal packages.